File: complex_dice_roller.py
import random as r

# prints instructions and takes user input
print("Input formatted as NdN, with constants being Nd1")
input_string = input("What would you like to roll? ") # I learned in Python2, so I would have used raw_input there

# preps for loop
input_arr = input_string.split()
sum = 0

for x in input_arr:
    nums = x.split('d') # splits on d; 5d6 becomes [5,6]

    for y in range(int(nums[0])): # first num is the number of rolls on this die
        roll = r.randrange(int(nums[1])) + 1 # second num is the size of the die
        sum = sum + roll
        print(roll)

print("Total: " + str(sum)) # returns

File: complex_dice_roller.cpp
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

using namespace std;

int main() {
  cout << "Input formatted as NdN, with constants being Nd1\nWhat would you like to roll?" << endl;
  
  string line;

  getline(cin, line);
  stringstream ss(line); // we'll read from here

  vector<string> input_arr;

  // loads vector for input processing
  while (ss >> line) {
    input_arr.push_back(line);
  }

  int sum = 0;

  for (string s : input_arr) {
    // Get substring from start to "d"
    string first = s.substr(0, s.find("d"));

    // Get substring from "d" to end
    string second = s.substr(s.find("d") + 1, s.length());

    // convert both to integers
    int num = stoi(first);
    int die = stoi(second);

    // roll num dice with sides in [1:die]
    for (int idx = 0; idx < num; idx++) {
      int roll = rand() % die + 1;
      sum += roll;

      // Uncomment to play Yahtzee
      //cout << roll << endl;
    }
  }

  // Final sum
  cout << sum << endl;
  return 0;
}
